libraries/AP_Networking/AP_Networking.cpp

```diff
@@ -0,0 +1,676 @@
+        _activeSettings.announce_ms = AP_HAL::millis();
+    }
+#endif
+
+    GCS_SEND_TEXT(MAV_SEVERITY_DEBUG,"NET: Initialized");
+    _init.done = true;
+}
+
+void AP_Networking::announce_address_changes()
+{
+    const uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - _activeSettings.announce_ms < 1000) {
+        // Never announce changes any faster than 1 sec
+        return;
+    }
+
+    const uint32_t ip = lwipGetIp();
+    const uint32_t nm = lwipGetNetmask();
+    const uint32_t gw = lwipGetGateway();
+
+    // struct netif thisif = lwipGetNetIf();
+    // const uint32_t ip = thisif.ip_addr.u_addr.ip4.addr;
+    // const uint32_t nm = thisif.netmask.u_addr.ip4.addr;
+    // const uint32_t gw = thisif.gw.u_addr.ip4.addr;
+
+
+    if (_activeSettings.announce_at_boot_done &&
+        ip == _activeSettings.ip &&
+        nm == _activeSettings.nm &&
+        gw == _activeSettings.gw)
+    {
+        // nothing changed and we've already printed it at least once. Nothing to do.
+        return;
+    }
+
+    _activeSettings.ip = ip;
+    _activeSettings.nm = nm;
+    _activeSettings.gw = gw;
+
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "NET: IP      %s", get_ip_active_str());
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "NET: Mask    %s", get_netmask_active_str());
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "NET: Gateway %s", get_gateway_active_str());
+
+    if (!_activeSettings.announce_at_boot_done && ip == 0 && nm == 0 && gw == 0 && get_dhcp_enabled()) {
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"NET: DHCP enabled, waiting for IP");
+    }
+
+    _activeSettings.announce_ms = now_ms;
+    _activeSettings.announce_at_boot_done = true;
+}
+
+void AP_Networking::update()
+{
+    if (!is_healthy()) {
+        return;
+    }
+
+    announce_address_changes();
+
+    for (uint8_t i=0; i<_num_instances; i++) {
+        if (_drivers[i] != nullptr && get_type(i) != AP_Networking_Params::Type::None) {
+            _drivers[i]->update();
+        }
+    }
+}
+
+AP_Networking_Params::Type AP_Networking::get_type(const uint8_t instance) const
+{
+    if (instance >= AP_NETWORKING_MAX_INSTANCES) {
+        return AP_Networking_Params::Type::None;
+    }
+    return (AP_Networking_Params::Type)_params[instance].type.get();
+}
+
+uint32_t AP_Networking::convert_netmask_bitcount_to_ip(const uint32_t netmask_bitcount)
+{
+    if (netmask_bitcount > 32) {
+        return 0;
+    }
+
+    uint32_t netmask_ip = 0;
+    for (uint32_t i=0; i<netmask_bitcount; i++) {
+        netmask_ip |= (1UL << i);
+    }
+    return netmask_ip;
+}
+
+uint8_t AP_Networking::convert_netmask_ip_to_bitcount(const uint32_t netmask_ip)
+{
+    uint32_t netmask_bitcount = 0;
+    for (uint32_t i=0; i<32; i++) {
+        // note, netmask LSB is IP MSB
+        if ((netmask_ip & (1UL<<i)) == 0) {
+            break;
+        }
+        netmask_bitcount++;
+    }
+    return netmask_bitcount;
+}
+
+uint32_t AP_Networking::convert_str_to_ip(char* ip_str)
```

Comment:
Since ip_str doesn't change, want to make it const? 

[inet_pton](https://man7.org/linux/man-pages/man3/inet_pton.3.html) takes in a const `src` param.